src/control/controlvalue.h

```diff
@@ -152,13 +153,10 @@ class ControlValueAtomicBase<T, cRingSize, true> {
 #endif
 };
 
-// ControlValueAtomic is a wrapper around ControlValueAtomicBase which uses the
-// sizeof(T) to determine which underlying implementation of
-// ControlValueAtomicBase to use. For types where sizeof(T) <= sizeof(void*),
-// the specialized implementation of ControlValueAtomicBase for types that are
-// atomic on the architecture is used.
-template <typename T, int cRingSize = kDefaultRingSize>
-class ControlValueAtomic : public ControlValueAtomicBase<T, cRingSize, sizeof(T) <= sizeof(void*)> {
+template<typename T, int cRingSize = kDefaultRingSize>
+class ControlValueAtomic : public ControlValueAtomicBase<T,
+                                   cRingSize,
+                                   std::atomic<T>::is_always_lock_free> {
```

Comment:
We are now depending on the std library implementation. Can you add a static assert that confirms our assumption, just in case? 
`static_assert((std::atomic<T>::is_always_lock_free == true) || (sizeof(T) > sizeof(void*))`

Reply:
I'm afraid I can't quite follow you. The static assert you proposed doesn't make much sense to me logically because we if its not `always_lock_free`, why would it be when `T` is larger than `void*`?

Reply:
Let's imagine an incomplete library implementation where a certain type is not flagged as "is_always_lock_free" even though it is smaller than void* and a solution should be possible. 
In this case the first part is false and the assertion is violated if the type is smaller than "void*". In such a case the old code would have used the aligned value and the new implementation uses the ring.

Reply:
right, but on that same hypothetical architecture where types smaller or equal than `void*` were not lock free, the assumption that aligned access is always atomic may be just as wrong, in which case we do want to use the ring. Besides, I don't think such an architecture exists right now (at least it does not seem to be the case on x86, arm or RISC-V).

If we want a solution that still performs optimally at runtime depending on the target, we could create a wrapper that switches its type depending on the runtime [`std::atomic<T>::is_lock_free()`](https://en.cppreference.com/w/cpp/atomic/atomic/is_lock_free) function. 

In general I don't think we should mistrust the standard library because the probability that it is incomplete is much lower than the probability that we make a mistake and the additional complexity is not worth it. The established std implementations are used and tested by so many people and companies that have a close eye on its completeness and correctness that its fine to rely on that.

Reply:
> right, but on that same hypothetical architecture where types smaller or equal than void* were not lock free, the assumption that aligned access is always atomic may be just as wrong,

Yes. In such a case the assertion is violated for a good reason and we can fix it by a `reinterpret_cast` to a lock free type or using a memory region where atomic access is possible. 

> If we want a solution that still performs optimally at runtime depending on the target, we could create a wrapper that switches its type depending on the runtime [std::atomic<T>::is_lock_free()](https://en.cppreference.com/w/cpp/atomic/atomic/is_lock_free) function.

Yes, the proposed assertion gives us just the evidence that there is such a problem to solve. 

> In general I don't think we should mistrust the standard library 

Better same than sorry. At least you find some related discussions: 
https://stackoverflow.com/questions/30048533/why-isnt-atomic-double-fully-implemented
https://github.com/WebAssembly/threads/issues/7

We are discussion here about a single compile time check, that is impossible to trigger from our current knowledge.

Reply:
sure I can add that check so we get alerted to such a case, but I don't think we should leverage implicit assumptions to fix them (the reinterpret_cast or manually layout out the memory) and instead use std-library tools. I'll add the assertion.

Reply:
done